### DISCUSSION TITLE
Using std::result::Result in request/response structures

### DISCUSSION THREAD

**Discussion by Slonegg** [juhaku/utoipa#767] [Q&A] [answered]
I have a structure that contains `Result`. Simplified example:
```
#[derive(Serialize, Deserialize, ToSchema)]
pub struct CommandResult {
    result: Result<CommandOutput, String>,
}
```

The endpoint works fine. However, I'm struggling to get the documentation rendered with Utoipa.
![image](https://github.com/juhaku/utoipa/assets/2404864/4ca5bc62-d0ea-4040-a2a5-dca9c86882d2)

Apparently the problem is with the schema for the `Result` type. Adding `Result<CommandOutput, String>` to schemas results in the following error:
```
 the trait `ToSchema<'_>` is not implemented for `std::result::Result<CommandOutput, std::string::String>`
```

Or is the idea to implement `ToSchema` for the aforementioned type somehow?

**Comment by juhaku** [comment]
Hey,

This is same situation as with serde in order to support external types that are not declared with `Serialize` or `Deserialize`. To this purpose serde has the `serialize_with` attribute.  For what comes to `utoipa` there are few ways around it depending the needs of user in general. You might want to follow links here https://github.com/juhaku/utoipa/issues/656 and https://github.com/juhaku/utoipa/issues/507#issuecomment-1465007714.

But in nutshell you need to manually create the schema for the the result type which should mimic the what the Result should look like in generated OpenAPI. Most likely it should be something like `oneOf` selection.

Hope this helps.

**Comment by Slonegg** [comment, marked as answer]
Thank you. Adding the following struct has resolved the issue:
```
#[derive(ToSchema)]
#[schema(as = Result<CommandOutput, String>)]
pub enum CommandResultSchema {
    Ok(CommandOutput),
    Err(String),
}
```
